javalib/src/main/scala/java/lang/String.scala

```diff
@@ -1,42 +1,908 @@
-  def substring(beginIndex: scala.Int): _String = ???
-  def substring(beginIndex: scala.Int, endIndex: scala.Int): _String = ???
-  def startsWith(prefix: _String): scala.Boolean = ???
-  def toLowerCase(): _String = ???
-  def toCharArray(): Array[Char] = ???
-  def equalsIgnoreCase(str: _String): scala.Boolean = ???
-  def getChars(srcBegin: scala.Int, srcEnd: scala.Int, dst: Array[scala.Char],
-      dstBegin: scala.Int): Unit = ???
-  override def hashCode(): scala.Int = ???
-
-  // TODO: rest of the api
+import java.io.{Serializable, UnsupportedEncodingException}
+import java.util._
+import java.util.regex._
+import java.nio._
+import java.nio.charset._
+
+import scala.util.Try
+
+final class _String private ()
+    extends Serializable with Comparable[_String] with CharSequence {
+  private var value         = new Array[Char](0)
+  private var offset        = 0
+  private var count         = 0
+  private var hashcode: Int = _
+
+  def this(data: Array[scala.Byte], start: Int, length: Int) = {
+    this()
+    if (start >= 0 && 0 <= length && length <= data.length - start) {
+      offset = 0
+      val charset    = defaultCharset()
+      val charBuffer = charset.decode(ByteBuffer.wrap(data, start, length))
+      if (charBuffer.length > 0) {
+        value = charBuffer.array
+        count = charBuffer.length
+      } else {
+        count = 0
+        value = new Array[Char](0)
+      }
+    } else {
+      throw new StringIndexOutOfBoundsException()
+    }
+  }
+
+  def this(data: Array[scala.Byte], _high: Int, start: Int, length: Int) = {
+    this()
+    if (length <= data.length - start && start >= 0 && 0 <= length) {
+      offset = 0
+      value = new Array[Char](length)
+      var high = _high
+      high <<= 8
+      for (i <- 0 until length) {
+        value(i) = (high + data(start + i) & 0xff).toChar
+      }
+    } else {
+      throw new StringIndexOutOfBoundsException()
+    }
+  }
+
+  def this(
+      data: Array[scala.Byte], start: Int, length: Int, encoding: _String) = {
+    this()
+    if (encoding == null) {
+      throw new NullPointerException()
+    } else {
+      if (start >= 0 && 0 <= length && length <= data.length - start) {
+        offset = 0
+        val charset = getCharset(encoding)
+        val charBuffer = Try(
+            charset.decode(ByteBuffer.wrap(data, start, length))
+        ).getOrElse(CharBuffer.wrap("?".toCharArray))
+
+        if (charBuffer.length > 0) {
+          value = charBuffer.array
+          count = charBuffer.length
+        } else {
+          count = 0
+          value = new Array[Char](0)
+        }
+      } else {
+        throw new StringIndexOutOfBoundsException()
+      }
+    }
+  }
+
+  def this(
+      data: Array[scala.Byte], start: Int, length: Int, encoding: Charset) = {
+    this()
+    if (encoding == null) {
+      throw new NullPointerException()
+    } else {
+      if (start >= 0 && 0 <= length && length <= data.length - start) {
```

Comment:
These checks are redundant, because they are supposed to be performed by `ByteBuffer.wrap`.